File: svm/svm.go
// svm.go
//
// Support vector machine implementation, based mainly on:
// https://towardsdatascience.com/svm-implementation-from-scratch-python-2db2fc52e5c2
//
// Also useful:
// https://www.mltut.com/svm-implementation-in-python-from-scratch/
// https://towardsdatascience.com/the-kernel-trick-c98cdbcaeb3f
// https://www.kaggle.com/code/prathameshbhalekar/svm-with-kernel-trick-from-scratch
//
// AK, 26/01/2023

package svm

import (
	"fmt"
	"math"

	"gonum.org/v1/gonum/mat"
)

// Global hyperparameters
var maxIterations int = 5000
var regularizationStrength float64 = 10000
var learningRate float64 = 0.000001
var costThreshold float64 = 0.01 // stop when improvement less than this fraction

// Train a Support Vector Machine model, using stochastic gradient descent.
// Returns a vector of weights that can be used to predict.
func sgd(X, Y *mat.Dense) *mat.VecDense {

	// Initialize weights as a vector of zeros
	nr, nc := X.Dims()
	W := mat.NewVecDense(nc, nil)
	fmt.Printf("Initial cost = %f\n", computeCost(W, X, Y))

	// Iterate until no more improvement, or maximum iterations
	prevCost := math.MaxFloat64
	for iter := 1; iter <= maxIterations; iter++ {

		// Do each row, keep adjusting weights
		// TODO: X, Y = shuffle(features, outputs)
		for i := 0; i < nr; i++ {

			// Get gradient for this row
			x := X.RowView(i) // mat.Vector
			y := Y.At(i, 0)   // float64
			ascent := calculateCostGradient(W, &x, y)

			// Python: W = W - (learningRate * ascent)
			ascent.ScaleVec(learningRate, ascent)
			W.SubVec(W, ascent)
		}

		// Stop when converged, i.e., no more improvement
		cost := computeCost(W, X, Y)
		fmt.Printf("Iteration %d: cost = %f\n", iter, cost)
		if math.Abs(prevCost-cost) < costThreshold*prevCost {
			break
		}
		prevCost = cost
	}

	// Return final weights
	return W
}

// Compute cost gradient for training SVM
// Assumes X and W are vectors (one row), Y is one number
func calculateCostGradient(W *mat.VecDense, x *mat.Vector, y float64) *mat.VecDense {

	// Calculate total distance
	// Python: d = 1 - (Y * np.sum(X * W))
	// was: distance = 1 - (Y_batch * np.dot(X_batch, W))
	nc := (*x).Len()
	var dist float64
	for i := 0; i < nc; i++ { // equivalent to: Y * np.sum(X * W)
		dist += W.AtVec(i) * (*x).AtVec(i) * y
	}
	if dist < 1 {
		dist = 1 - dist
	} else {
		dist = 0
	}

	// Calculate dw
	// dw = np.zeros(len(W))
	// if dist > 0:
	//     dw = W - (regularization_strength * Y * X)
	dw := mat.NewVecDense(nc, nil) // 31 x 1
	if dist > 0 {                  // dist no longer used!
		for i := 0; i < nc; i++ {
			dw.SetVec(i, W.AtVec(i)-regularizationStrength*y*(*x).AtVec(i))
		}
	}

	// Return a vector of weight differences
	return dw
}

// Compute cost for SVM
func computeCost(W *mat.VecDense, X, Y *mat.Dense) float64 {

	// Calculate distances
	// Python: distances = 1 - Y * np.dot(X, W)
	// distances[distances < 0] = 0  # i.e., max(0, distance)
	nr, nc := X.Dims()
	dist := mat.NewDense(nr, 1, nil)
	dist.Mul(X, W)
	dist.MulElem(Y, dist)
	dist.Apply(func(i, j int, v float64) float64 {
		if v < 1 {
			return 1 - v
		} else {
			return 0
		}
	}, dist)

	// Python: sumDistances = np.sum(distances)
	var sumDist float64
	for i := 0; i < nr; i++ {
		sumDist += dist.At(i, 0)
	}

	// Calculate dot(W, W)
	var cost float64 = 0
	for c := 0; c < nc; c++ {
		w := W.AtVec(c)
		cost += w * w
	}

	// Calculate cost
	// Python: cost = 1 / 2 * np.dot(W, W) + hinge_loss
	hingeLoss := regularizationStrength * (sumDist / float64(nr))
	return cost/2 + hingeLoss
}

File: svm/svm.py
import numpy as np
import pandas as pd
import statsmodels.api as sm
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split as tts
from sklearn.metrics import accuracy_score, recall_score, precision_score
from sklearn.utils import shuffle


# Feature selection: remove correlated features
def remove_correlated_features(X):
    corr_threshold = 0.9
    corr = X.corr()
    drop_columns = np.full(corr.shape[0], False, dtype=bool)
    for i in range(corr.shape[0]):
        for j in range(i + 1, corr.shape[0]):
            if corr.iloc[i, j] >= corr_threshold:
                drop_columns[j] = True
    columns_dropped = X.columns[drop_columns]
    X.drop(columns_dropped, axis=1, inplace=True)
    return columns_dropped

# Remove less significant features
def remove_less_significant_features(X, Y):
    sl = 0.05
    regression_ols = None
    columns_dropped = np.array([])
    for itr in range(0, len(X.columns)):
        regression_ols = sm.OLS(Y, X).fit()
        max_col = regression_ols.pvalues.idxmax()
        max_val = regression_ols.pvalues.max()
        if max_val > sl:
            X.drop(max_col, axis='columns', inplace=True)
            columns_dropped = np.append(columns_dropped, [max_col])
        else:
            break
    regression_ols.summary()
    return columns_dropped

# Compute cost, for model training
def compute_cost(W, X, Y):

    # Calculate distances
    distances = 1 - Y * np.dot(X, W)
    distances[distances < 0] = 0  # i.e., max(0, distance)
    print('***', W.shape, X.shape, Y.shape, distances.shape)
    
    # calculate hinge loss
    N = X.shape[0]
    hinge_loss = regularization_strength * (np.sum(distances) / N)

    # calculate cost
    cost = 1 / 2 * np.dot(W, W) + hinge_loss
    return cost


# I haven't tested it but this same function should work for
# vanilla and mini-batch gradient descent as well
def calculate_cost_gradient(W, X_batch, Y_batch):

    print('W:', W, W.shape)
    print('X:', X_batch, X_batch.shape)
    print('Y:', Y_batch, Y_batch.shape)

    # if only one example is passed (eg. in case of SGD)
    if type(Y_batch) == np.float64:
        #print('Converting to arrays:', X_batch.shape, Y_batch.shape)
        Y_batch = np.array([Y_batch])
        X_batch = np.array([X_batch])  # gives multidimensional array
        #print('  =>', X_batch.shape, Y_batch.shape)

    distance = 1 - (Y_batch * np.dot(X_batch, W))
    dw = np.zeros(len(W))

    for ind, d in enumerate(distance):
        if max(0, d) == 0:
            di = W
        else:
            di = W - (regularization_strength * Y_batch[ind] * X_batch[ind])
        dw += di

    dw = dw/len(Y_batch)  # average
    return dw


# Stochastic Gradient Descent
def sgd(features, outputs):

    max_epochs = 5000
    weights = np.zeros(features.shape[1]) # row of 13 zeros
    prev_cost = float("inf")
    cost_threshold = 0.01  # in percent

    print("Initial cost:", compute_cost(weights, features, outputs))
        
    # stochastic gradient descent
    for epoch in range(1, max_epochs):

        # TODO: Shuffle to prevent repeating update cycles (but lowers 
        # performance a bit)
        # X:(455, 13)
        #X, Y = shuffle(features, outputs)
        X, Y = features, outputs

        # Process each row of features, ascent is a vector of 13 numbers
        for ind, x in enumerate(X): # each row number, row data
            ascent = calculate_cost_gradient(weights, x, Y[ind])
            weights = weights - (learning_rate * ascent)

        # Stop when no longer improving
        cost = compute_cost(weights, features, outputs)
        print("Epoch %d: cost = %.2f" % (epoch, cost))
        if abs(prev_cost - cost) < cost_threshold * prev_cost:
            return weights
        prev_cost = cost

    return weights

# set hyper-parameters
regularization_strength = 10000
learning_rate = 0.000001

# read data in pandas (pd) data frame
print("reading dataset...")
data = pd.read_csv('../data/breastcancer.csv')

# drop last column (extra column added by pd)
# and unnecessary first column (id)
data.drop(data.columns[[-1, 0]], axis=1, inplace=True)

# convert categorical labels to numbers
print("Feature engineering...")
diag_map = {'M': 1.0, 'B': -1.0}
data['diagnosis'] = data['diagnosis'].map(diag_map)

# put features & outputs in different data frames
Y = data.loc[:, 'diagnosis']
X = data.iloc[:, 1:]

# TODO: filter features
#remove_correlated_features(X)
#remove_less_significant_features(X, Y)

# normalize data for better convergence and to prevent overflow
X_normalized = MinMaxScaler().fit_transform(X.values)
X = pd.DataFrame(X_normalized)

# insert 1 in every row for intercept b
# X: 569 rows x 13 cols (last is intercept 1)
# Y: vector of 569 values (1/-1)
X.insert(loc=len(X.columns), column='intercept', value=1)

# split data into train and test set
print("Splitting dataset into train and test sets...")
X_train, X_test, y_train, y_test = tts(X, Y, test_size=0.2, random_state=42)

# train the model
print("Training ...")
W = sgd(X_train.to_numpy(), y_train.to_numpy())
print("Weights:", W)

# test the model
print("Testing ...")
y_train_predicted = np.array([])
for i in range(X_train.shape[0]):
    yp = np.sign(np.dot(X_train.to_numpy()[i], W))
    y_train_predicted = np.append(y_train_predicted, yp)

y_test_predicted = np.array([])
for i in range(X_test.shape[0]):
    yp = np.sign(np.dot(X_test.to_numpy()[i], W))
    y_test_predicted = np.append(y_test_predicted, yp)

print("Accuracy on test dataset:", accuracy_score(y_test, y_test_predicted))
print("Recall on test dataset:", recall_score(y_test, y_test_predicted))
print("Precision on test dataset:", recall_score(y_test, y_test_predicted))

